File: src/services/ripple-assembler/macro-formatter.ts
import type { Instruction } from './types.ts';
import { Opcode, Register } from './types.ts';

export class MacroFormatter {
  private opcodeToMacro: Map<number, string> = new Map([
    [Opcode.NOP, '@OP_NOP'],
    [Opcode.ADD, '@OP_ADD'],
    [Opcode.SUB, '@OP_SUB'],
    [Opcode.AND, '@OP_AND'],
    [Opcode.OR, '@OP_OR'],
    [Opcode.XOR, '@OP_XOR'],
    [Opcode.SLL, '@OP_SLL'],
    [Opcode.SRL, '@OP_SRL'],
    [Opcode.SLT, '@OP_SLT'],
    [Opcode.SLTU, '@OP_SLTU'],
    [Opcode.ADDI, '@OP_ADDI'],
    [Opcode.ANDI, '@OP_ANDI'],
    [Opcode.ORI, '@OP_ORI'],
    [Opcode.XORI, '@OP_XORI'],
    [Opcode.LI, '@OP_LI'],
    [Opcode.SLLI, '@OP_SLLI'],
    [Opcode.SRLI, '@OP_SRLI'],
    [Opcode.LOAD, '@OP_LOAD'],
    [Opcode.STORE, '@OP_STOR'],
    [Opcode.JAL, '@OP_JAL'],
    [Opcode.JALR, '@OP_JALR'],
    [Opcode.BEQ, '@OP_BEQ'],
    [Opcode.BNE, '@OP_BNE'],
    [Opcode.BLT, '@OP_BLT'],
    [Opcode.BGE, '@OP_BGE'],
    [Opcode.BRK, '@OP_BRK'],
    [Opcode.MUL, '@OP_MUL'],
    [Opcode.DIV, '@OP_DIV'],
    [Opcode.MOD, '@OP_MOD'],
    [Opcode.MULI, '@OP_MULI'],
    [Opcode.DIVI, '@OP_DIVI'],
    [Opcode.MODI, '@OP_MODI'],
  ]);

  private registerToMacro(reg: number): string {
    switch (reg) {
      case 0:
        return '#R0'; // Zero register
      case 1:
        return '#PC'; // Program Counter
      case 2:
        return '#PCB'; // Program Counter Bank
      case 3:
        return '#RA'; // Return Address
      case 4:
        return '#RAB'; // Return Address Bank
      case 5:
        return '#RV0'; // Return Value 0
      case 6:
        return '#RV1'; // Return Value 1
      case 7:
        return '#A0'; // Argument 0
      case 8:
        return '#A1'; // Argument 1
      case 9:
        return '#A2'; // Argument 2
      case 10:
        return '#A3'; // Argument 3
      case 11:
        return '#X0'; // Reserved/Extended 0
      case 12:
        return '#X1'; // Reserved/Extended 1
      case 13:
        return '#X2'; // Reserved/Extended 2
      case 14:
        return '#X3'; // Reserved/Extended 3
      case 15:
        return '#T0'; // Temporary 0
      case 16:
        return '#T1'; // Temporary 1
      case 17:
        return '#T2'; // Temporary 2
      case 18:
        return '#T3'; // Temporary 3
      case 19:
        return '#T4'; // Temporary 4
      case 20:
        return '#T5'; // Temporary 5
      case 21:
        return '#T6'; // Temporary 6
      case 22:
        return '#T7'; // Temporary 7
      case 23:
        return '#S0'; // Saved 0
      case 24:
        return '#S1'; // Saved 1
      case 25:
        return '#S2'; // Saved 2
      case 26:
        return '#S3'; // Saved 3
      case 27:
        return '#SC'; // Allocator Scratch
      case 28:
        return '#SB'; // Stack Bank
      case 29:
        return '#SP'; // Stack Pointer
      case 30:
        return '#FP'; // Frame Pointer
      case 31:
        return '#GP'; // Global Pointer
      default:
        throw new Error(
          `Invalid register number: ${reg}. Valid registers are 0-31.`,
        );
    }
  }

  formatInstruction(
    instruction: Instruction,
    isFirst: boolean = false,
  ): string {
    // Debug LI instructions
    if (instruction.opcode === 0x0e) {
      console.log('Formatting LI instruction:', {
        opcode: instruction.opcode,
        word0: instruction.word0,
        word1: instruction.word1,
        word2: instruction.word2,
        word3: instruction.word3,
      });
    }

    let opcodeMacro = this.opcodeToMacro.get(instruction.opcode) || '@OP_NOP';

    // Special case for HALT (encoded as NOP 0,0,0)
    if (
      instruction.opcode === Opcode.NOP &&
      instruction.word1 === 0 &&
      instruction.word2 === 0 &&
      instruction.word3 === 0
    ) {
      opcodeMacro = '@OP_HALT';
    }

    // Special case for BRK (encoded as BRK 0,0,0)
    if (
      instruction.opcode === Opcode.BRK &&
      instruction.word1 === 0 &&
      instruction.word2 === 0 &&
      instruction.word3 === 0
    ) {
      opcodeMacro = '@OP_BRK';
    }

    const formatOperand = (
      value: number,
      isRegister: boolean = false,
    ): string => {
      if (isRegister) {
        return this.registerToMacro(value);
      }
      return value.toString();
    };

    // Determine which operands are registers based on opcode
    const isRFormat = [
      Opcode.ADD,
      Opcode.SUB,
      Opcode.AND,
      Opcode.OR,
      Opcode.XOR,
      Opcode.SLL,
      Opcode.SRL,
      Opcode.SLT,
      Opcode.SLTU,
      Opcode.JALR,
      Opcode.MUL,
      Opcode.DIV,
      Opcode.MOD,
    ].includes(instruction.opcode);

    const isIFormat = [
      Opcode.ADDI,
      Opcode.ANDI,
      Opcode.ORI,
      Opcode.XORI,
      Opcode.SLLI,
      Opcode.SRLI,
      Opcode.LOAD,
      Opcode.STORE,
      Opcode.BEQ,
      Opcode.BNE,
      Opcode.BLT,
      Opcode.BGE,
      Opcode.MULI,
      Opcode.DIVI,
      Opcode.MODI,
    ].includes(instruction.opcode);

    let word1Str: string;
    let word2Str: string;
    let word3Str: string;

    if (isRFormat) {
      word1Str = formatOperand(instruction.word1, true);
      word2Str = formatOperand(instruction.word2, true);
      word3Str = formatOperand(instruction.word3, true);
    } else if (isIFormat) {
      // Special case for LOAD: all positions can be registers
      if (instruction.opcode === Opcode.LOAD) {
        word1Str = formatOperand(instruction.word1, true); // rd is register
        word2Str = formatOperand(instruction.word2, true); // bank is register (or 0)
        word3Str = formatOperand(instruction.word3, true); // addr is register
      } else {
        word1Str = formatOperand(instruction.word1, true);
        word2Str = formatOperand(instruction.word2, true);
        word3Str = formatOperand(instruction.word3, false);
      }
    } else if (instruction.opcode === Opcode.LI) {
      word1Str = formatOperand(instruction.word1, true);
      word2Str = formatOperand(instruction.word2, false);
      word3Str = formatOperand(instruction.word3, false);
    } else if (instruction.opcode === Opcode.JAL) {
      // JAL uses address in word3
      word1Str = formatOperand(instruction.word1, false);
      word2Str = formatOperand(instruction.word2, false);
      word3Str = formatOperand(instruction.word3, false);
    } else {
      word1Str = formatOperand(instruction.word1, false);
      word2Str = formatOperand(instruction.word2, false);
      word3Str = formatOperand(instruction.word3, false);
    }

    const cmdType = isFirst ? '@program_start' : '@cmd';

    return `${cmdType}(${opcodeMacro.padEnd(10)}, ${word1Str.padEnd(4)}, ${word2Str.padEnd(4)}, ${word3Str})`;
  }

  formatProgram(
    instructions: Instruction[],
    comments?: Map<number, string>,
  ): string {
    const lines: string[] = [];

    instructions.forEach((instruction, index) => {
      const comment = comments?.get(index);
      const formattedInst = this.formatInstruction(instruction, index === 0);

      if (comment) {
        lines.push(`${formattedInst}        // ${comment}`);
      } else {
        lines.push(formattedInst);
      }
    });

    lines.push('@program_end');

    return lines.join('\n');
  }

  formatDataSection(data: number[]): string {
    const lines: string[] = [];

    lines.push(`// Data segment`);
    lines.push(`@lane(#L_MEM,`);

    // Convert data array to mixed format array (decimal, hex, or char)
    const formattedData: string[] = data.map((value) => {
      // Use character format for printable ASCII (space through ~)
      if (value >= 32 && value <= 126) {
        const char = String.fromCharCode(value);
        // Escape single quotes and backslashes
        if (char === "'") {
          return "'\\''";
        } else if (char === '\\') {
          return "'\\\\'";
        }
        return `'${char}'`;
      }
      // Use hex for values that look better in hex
      else if (value >= 128 || (value > 15 && value % 16 === 0)) {
        return `0x${value.toString(16).toUpperCase()}`;
      }
      // Use decimal for everything else
      else {
        return value.toString();
      }
    });

    // Create the {for} loop with formatted data
    lines.push(`  {for(s in {${formattedData.join(',')}}, @set(s) @nextword)}`);

    lines.push(`)`);

    return lines.join('\n');
  }

  formatFullProgram(
    instructions: Instruction[],
    data?: number[],
    comments?: Map<number, string>,
    header?: string,
  ): string {
    const lines: string[] = [];

    if (header) {
      lines.push(`// ${header}`);
      lines.push('');
    }

    lines.push('@prg(');

    // Format data section or @nop if empty
    if (data && data.length > 0) {
      lines.push('  // Memory');
      const dataLines = this.formatDataSection(data).split('\n');
      dataLines.forEach((line, index) => {
        if (index === 0 && line.startsWith('// Data segment')) {
          // Skip the old header comment
          return;
        }
        if (line === ')') {
          // Change the closing paren to include comma
          lines.push('  ),');
        } else {
          lines.push('  ' + line);
        }
      });
      lines.push('  ');
    } else {
      lines.push('  @nop,');
      lines.push('  ');
    }

    // Format program section or @nop if empty
    if (instructions && instructions.length > 0) {
      lines.push('  // Program');
      const programLines = this.formatProgram(instructions, comments).split(
        '\n',
      );
      programLines.forEach((line) => {
        lines.push('  ' + line);
      });
    } else {
      lines.push('  @nop');
    }

    lines.push(')');

    return lines.join('\n');
  }
}

File: src/ripple-asm/src/macro_formatter.rs
use crate::types::{Instruction, Opcode, Register};
use crate::linker::LinkedProgram;
use std::collections::HashMap;

const CPU_HEADER: &str = include_str!("cpu_header.bfm");
const CPU_FOOTER: &str = include_str!("cpu_footer.bfm");

pub struct MacroFormatter {
    opcode_to_macro: HashMap<u8, &'static str>,
}

impl MacroFormatter {
    pub fn new() -> Self {
        let mut opcode_to_macro = HashMap::new();
        
        opcode_to_macro.insert(Opcode::Nop as u8, "@OP_NOP");
        opcode_to_macro.insert(Opcode::Add as u8, "@OP_ADD");
        opcode_to_macro.insert(Opcode::Sub as u8, "@OP_SUB");
        opcode_to_macro.insert(Opcode::And as u8, "@OP_AND");
        opcode_to_macro.insert(Opcode::Or as u8, "@OP_OR");
        opcode_to_macro.insert(Opcode::Xor as u8, "@OP_XOR");
        opcode_to_macro.insert(Opcode::Sll as u8, "@OP_SLL");
        opcode_to_macro.insert(Opcode::Srl as u8, "@OP_SRL");
        opcode_to_macro.insert(Opcode::Slt as u8, "@OP_SLT");
        opcode_to_macro.insert(Opcode::Sltu as u8, "@OP_SLTU");
        opcode_to_macro.insert(Opcode::Addi as u8, "@OP_ADDI");
        opcode_to_macro.insert(Opcode::Andi as u8, "@OP_ANDI");
        opcode_to_macro.insert(Opcode::Ori as u8, "@OP_ORI");
        opcode_to_macro.insert(Opcode::Xori as u8, "@OP_XORI");
        opcode_to_macro.insert(Opcode::Li as u8, "@OP_LI");
        opcode_to_macro.insert(Opcode::Slli as u8, "@OP_SLLI");
        opcode_to_macro.insert(Opcode::Srli as u8, "@OP_SRLI");
        opcode_to_macro.insert(Opcode::Load as u8, "@OP_LOAD");
        opcode_to_macro.insert(Opcode::Store as u8, "@OP_STOR");
        opcode_to_macro.insert(Opcode::Jal as u8, "@OP_JAL");
        opcode_to_macro.insert(Opcode::Jalr as u8, "@OP_JALR");
        opcode_to_macro.insert(Opcode::Beq as u8, "@OP_BEQ");
        opcode_to_macro.insert(Opcode::Bne as u8, "@OP_BNE");
        opcode_to_macro.insert(Opcode::Blt as u8, "@OP_BLT");
        opcode_to_macro.insert(Opcode::Bge as u8, "@OP_BGE");
        opcode_to_macro.insert(Opcode::Brk as u8, "@OP_BRK");
        opcode_to_macro.insert(Opcode::Mul as u8, "@OP_MUL");
        opcode_to_macro.insert(Opcode::Div as u8, "@OP_DIV");
        opcode_to_macro.insert(Opcode::Mod as u8, "@OP_MOD");
        opcode_to_macro.insert(Opcode::Muli as u8, "@OP_MULI");
        opcode_to_macro.insert(Opcode::Divi as u8, "@OP_DIVI");
        opcode_to_macro.insert(Opcode::Modi as u8, "@OP_MODI");
        
        Self { opcode_to_macro }
    }

    fn register_to_macro(&self, reg: u16) -> String {
        match reg as u8 {
            r if r == Register::R0 as u8 => "@R0".to_string(),
            r if r == Register::Pc as u8 => "@PC".to_string(),
            r if r == Register::Pcb as u8 => "@PCB".to_string(),
            r if r == Register::Ra as u8 => "@RA".to_string(),
            r if r == Register::Rab as u8 => "@RAB".to_string(),
            r if r >= 5 && r <= 17 => format!("@R{}", r - 2),
            r => format!("#R{}", r),
        }
    }

    pub fn format_instruction(&self, instruction: &Instruction, is_first: bool) -> String {
        let mut opcode_macro = self.opcode_to_macro
            .get(&instruction.opcode)
            .copied()
            .unwrap_or_else(|| {
                panic!("Unknown opcode {} (0x{:02x}) in macro formatter", 
                       instruction.opcode, instruction.opcode)
            });
        
        // Special case for HALT (encoded as NOP 0,0,0)
        if instruction.is_halt() {
            opcode_macro = "@OP_HALT";
        }
        
        // Special case for BRK (encoded as BRK 0,0,0)
        if instruction.opcode == Opcode::Brk as u8 && 
            instruction.word1 == 0 && 
            instruction.word2 == 0 && 
            instruction.word3 == 0 {
            opcode_macro = "@OP_BRK";
        }

        let opcode = instruction.opcode;
        let format_operand = |value: u16, is_register: bool| -> String {
            if is_register {
                self.register_to_macro(value)
            } else {
                value.to_string()
            }
        };

        // Determine which operands are registers based on opcode
        let is_r_format = matches!(opcode,
            x if x == Opcode::Add as u8 || x == Opcode::Sub as u8 || 
                 x == Opcode::And as u8 || x == Opcode::Or as u8 || 
                 x == Opcode::Xor as u8 || x == Opcode::Sll as u8 || 
                 x == Opcode::Srl as u8 || x == Opcode::Slt as u8 || 
                 x == Opcode::Sltu as u8 || x == Opcode::Jalr as u8 ||
                 x == Opcode::Mul as u8 || x == Opcode::Div as u8 ||
                 x == Opcode::Mod as u8
        );

        let is_i_format = matches!(opcode,
            x if x == Opcode::Addi as u8 || x == Opcode::Andi as u8 || 
                 x == Opcode::Ori as u8 || x == Opcode::Xori as u8 ||
                 x == Opcode::Slli as u8 || x == Opcode::Srli as u8 || 
                 x == Opcode::Load as u8 || x == Opcode::Store as u8 ||
                 x == Opcode::Beq as u8 || x == Opcode::Bne as u8 || 
                 x == Opcode::Blt as u8 || x == Opcode::Bge as u8 ||
                 x == Opcode::Muli as u8 || x == Opcode::Divi as u8 ||
                 x == Opcode::Modi as u8
        );

        let (word1_str, word2_str, word3_str) = if is_r_format {
            (
                format_operand(instruction.word1, true),
                format_operand(instruction.word2, true),
                format_operand(instruction.word3, true),
            )
        } else if is_i_format {
            // Special case for LOAD: all positions can be registers
            if instruction.opcode == Opcode::Load as u8 {
                (
                    format_operand(instruction.word1, true),
                    format_operand(instruction.word2, true),
                    format_operand(instruction.word3, true),
                )
            } else {
                (
                    format_operand(instruction.word1, true),
                    format_operand(instruction.word2, true),
                    format_operand(instruction.word3, false),
                )
            }
        } else if instruction.opcode == Opcode::Li as u8 {
            (
                format_operand(instruction.word1, true),
                format_operand(instruction.word2, false),
                format_operand(instruction.word3, false),
            )
        } else if instruction.opcode == Opcode::Jal as u8 {
            (
                format_operand(instruction.word1, false),
                format_operand(instruction.word2, false),
                format_operand(instruction.word3, false),
            )
        } else {
            (
                format_operand(instruction.word1, false),
                format_operand(instruction.word2, false),
                format_operand(instruction.word3, false),
            )
        };

        let cmd_type = if is_first { "@program_start" } else { "@cmd" };
        
        format!(
            "{}({:10}, {:4}, {:4}, {})",
            cmd_type,
            opcode_macro,
            word1_str,
            word2_str,
            word3_str
        )
    }

    pub fn format_program(&self, instructions: &[Instruction], comments: Option<&HashMap<usize, String>>) -> String {
        let mut lines = Vec::new();
        
        for (index, instruction) in instructions.iter().enumerate() {
            let comment = comments.and_then(|c| c.get(&index));
            let formatted_inst = self.format_instruction(instruction, index == 0);
            
            if let Some(comment_text) = comment {
                lines.push(format!("{}        // {}", formatted_inst, comment_text));
            } else {
                lines.push(formatted_inst);
            }
        }
        
        lines.push("@program_end".to_string());
        
        lines.join("\n")
    }

    pub fn format_data_section(&self, data: &[u8]) -> String {
        let mut lines = Vec::new();
        
        lines.push("// Data segment".to_string());
        lines.push("@lane(#L_MEM,".to_string());
        
        // Convert data array to mixed format array (decimal, hex, or char)
        let formatted_data: Vec<String> = data.iter().map(|&value| {
            match value {
                // Special characters - use decimal
                9 => "9".to_string(),    // tab
                10 => "10".to_string(),  // newline
                13 => "13".to_string(),  // carriage return
                // Printable ASCII (space through ~)
                32..=126 => {
                    let ch = value as char;
                    // Escape single quotes and backslashes
                    match ch {
                        '\'' => "'\\'".to_string(),
                        '\\' => "'\\\\'".to_string(),
                        _ => format!("'{}'", ch),
                    }
                }
                // Use hex for high values
                128..=255 => format!("0x{:02X}", value),
                // Use decimal for everything else
                _ => value.to_string(),
            }
        }).collect();
        
        // Create the {for} loop with formatted data
        let data_str = formatted_data.join(",");
        lines.push(format!("  {{for(s in {{{}}}, @set(s) @nextword)}}", data_str));
        
        lines.push(")".to_string());
        
        lines.join("\n")
    }

    pub fn format_full_program(
        &self,
        instructions: &[Instruction],
        data: Option<&[u8]>,
        comments: Option<&HashMap<usize, String>>,
        header: Option<&str>,
    ) -> String {
        let mut lines = Vec::new();
        
        if let Some(header_text) = header {
            lines.push(format!("// {}", header_text));
            lines.push(String::new());
        }
        
        lines.push("@prg(".to_string());
        
        // Format data section or @nop if empty
        if let Some(data_bytes) = data {
            if !data_bytes.is_empty() {
                lines.push("  // Memory".to_string());
                let data_lines = self.format_data_section(data_bytes);
                for (idx, line) in data_lines.lines().enumerate() {
                    if idx == 0 && line.starts_with("// Data segment") {
                        // Skip the old header comment
                        continue;
                    }
                    if line == ")" {
                        // Change the closing paren to include comma
                        lines.push("  ),".to_string());
                    } else {
                        lines.push(format!("  {}", line));
                    }
                }
                lines.push("  ".to_string());
            } else {
                lines.push("  @nop,".to_string());
                lines.push("  ".to_string());
            }
        } else {
            lines.push("  @nop,".to_string());
            lines.push("  ".to_string());
        }
        
        // Format program section or @nop if empty
        if !instructions.is_empty() {
            lines.push("  // Program".to_string());
            let program_lines = self.format_program(instructions, comments);
            for line in program_lines.lines() {
                lines.push(format!("  {}", line));
            }
        } else {
            lines.push("  @nop".to_string());
        }
        
        lines.push(")".to_string());
        
        lines.join("\n")
    }

    pub fn format_linked_program(&self, program: &LinkedProgram) -> String {
        let comments = HashMap::new(); // Could be enhanced to preserve comments
        self.format_full_program(
            &program.instructions,
            Some(&program.data),
            Some(&comments),
            Some(&format!("Linked program with entry point at 0x{:08X}", program.entry_point)),
        )
    }

    pub fn format_linked_program_standalone(&self, program: &LinkedProgram, debug: bool) -> String {
        let mut output = String::new();
        
        // Add CPU header with modified DEBUG value
        let header = if debug {
            CPU_HEADER.replace("#define DEBUG 0", "#define DEBUG 1")
        } else {
            CPU_HEADER.to_string()
        };
        output.push_str(&header);
        output.push('\n');
        
        // Add the program content (without @prg wrapper since it's in the template)
        let comments = HashMap::new();
        let program_content = self.format_full_program(
            &program.instructions,
            Some(&program.data),
            Some(&comments),
            Some(&format!("Linked program with entry point at 0x{:08X}", program.entry_point)),
        );
        
        // Add indentation to match the template structure
        for line in program_content.lines() {
            output.push_str("  ");
            output.push_str(line);
            output.push('\n');
        }
        
        // Add CPU footer
        output.push_str(CPU_FOOTER);
        
        output
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_halt() {
        let formatter = MacroFormatter::new();
        let inst = Instruction::new(Opcode::Nop, 0, 0, 0);
        let formatted = formatter.format_instruction(&inst, true);
        
        assert!(formatted.contains("@OP_HALT"));
        assert!(formatted.starts_with("@program_start"));
    }

    #[test]
    fn test_format_add() {
        let formatter = MacroFormatter::new();
        let inst = Instruction::new(Opcode::Add, Register::R3 as u16, Register::R4 as u16, Register::R5 as u16);
        let formatted = formatter.format_instruction(&inst, false);
        
        assert!(formatted.contains("@OP_ADD"));
        assert!(formatted.contains("@R3"));
        assert!(formatted.contains("@R4"));
        assert!(formatted.contains("@R5"));
        assert!(formatted.starts_with("@cmd"));
    }

    #[test]
    fn test_format_data_section() {
        let formatter = MacroFormatter::new();
        let data = b"Hello\x00\x01\xFF";
        let formatted = formatter.format_data_section(data);
        
        assert!(formatted.contains("'H'"));
        assert!(formatted.contains("'e'"));
        assert!(formatted.contains("'l'"));
        assert!(formatted.contains("'o'"));
        assert!(formatted.contains("0"));  // null byte
        assert!(formatted.contains("1"));  // 0x01
        assert!(formatted.contains("0xFF"));  // 0xFF
    }

    #[test]
    fn test_format_full_program() {
        let formatter = MacroFormatter::new();
        let instructions = vec![
            Instruction::new(Opcode::Li, Register::R3 as u16, 42, 0),
            Instruction::new(Opcode::Nop, 0, 0, 0),
        ];
        let data = b"Test";
        
        let formatted = formatter.format_full_program(
            &instructions,
            Some(data),
            None,
            Some("Test Program"),
        );
        
        assert!(formatted.contains("// Test Program"));
        assert!(formatted.contains("@prg("));
        assert!(formatted.contains("@OP_LI"));
        assert!(formatted.contains("@OP_HALT"));
        assert!(formatted.contains("'T'"));
        assert!(formatted.contains("'e'"));
        assert!(formatted.contains("'s'"));
        assert!(formatted.contains("'t'"));
    }
}